### PR TITLE
Deprecate positional arguments for Citrine and Session

### PR DESCRIPTION
**Pull request by kroenlein**
# Citrine Python PR

## Description 
We've gotten signal that customers find the fact that the second positional argument to Citrine is `scheme` is confusing.  Having four valid positional arguments is also against [our coding-style](https://github.com/CitrineInformatics/citrine-python/blob/main/CONTRIBUTING.md#coding-style).  This deprecates having more than one argument (still allowing people to pass the API key).

Somehow, in the existing testing, the full signature was never actually exercised.  I've also added tests to address this.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [X] I have added tests for 100% coverage
- [X] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [X] I have bumped the version in setup.py

### CODE REVIEW

File: tests/test_citrine.py [hunk 1]
```diff
@@ -22,14 +21,37 @@ def refresh_token(expiration: datetime = None) -> dict:
...
 def test_citrine_creation():
-
     with requests_mock.Mocker() as m:
         m.post('https://citrine-testing.fake/api/v1/tokens/refresh', json=token_refresh_response)
         m.get('https://citrine-testing.fake/api/v1/utils/runtime-config', json=dict())
 
         assert '1234' == Citrine(api_key='1234', host='citrine-testing.fake').session.refresh_token
 
 
+def test_citrine_signature():
```
Comment (anoto-moniz) [inline tests/test_citrine.py]: This should probably include a test that we can provide `api_key` as a positional argument without raising a `DeprecationWarning`.

File: tests/test_session.py [hunk 2]
```diff
@@ -49,6 +49,32 @@ def session():
     return session
 
 
+def test_session_signature():
```
Comment (anoto-moniz) [inline tests/test_session.py]: Similar to above, but `refresh_token`.

File: src/citrine/_session.py [hunk 3]
```diff
@@ -32,11 +33,31 @@ class Session(requests.Session):
     """Wrapper around requests.Session that is both refresh-token and schema aware."""
 
     def __init__(self,
-                 refresh_token: str = environ.get('CITRINE_API_KEY'),
```
Comment (kroenlein) [inline src/citrine/_session.py:35]: When `environ.get` is in the method signature, the value is set at build time, which means we can't mock the value.

File: tests/test_citrine.py [hunk 4]
```diff
@@ -22,14 +21,57 @@ def refresh_token(expiration: datetime = None) -> dict:
...
 def test_citrine_creation():
-
     with requests_mock.Mocker() as m:
         m.post('https://citrine-testing.fake/api/v1/tokens/refresh', json=token_refresh_response)
         m.get('https://citrine-testing.fake/api/v1/utils/runtime-config', json=dict())
 
         assert '1234' == Citrine(api_key='1234', host='citrine-testing.fake').session.refresh_token
 
 
+def test_citrine_signature(monkeypatch):
```
Comment (anoto-moniz) [inline tests/test_citrine.py:31]: We should test that passing `api_key` as the only positional argument works.

File: tests/test_session.py [hunk 5]
```diff
@@ -49,6 +48,52 @@ def session():
     return session
 
 
+def test_session_signature(monkeypatch):
```
Comment (anoto-moniz) [inline tests/test_session.py:51]: We should test that passing `refresh_token` as the only positional argument works.

File: tests/test_session.py [hunk 6]
```diff
@@ -290,30 +335,3 @@ def test_patch(session: Session):
         m.patch('http://citrine-testing.fake/api/v1/bar/something', status_code=200, json=json_to_validate)
         response_json = session.patch_resource('bar/something', {"ignored": "true"})
         assert response_json == json_to_validate
-
-
-@mock.patch.object(Session, '_check_accounts_version')
-@mock.patch.object(Session, '_refresh_access_token')
-def test_base_url_assembly(*_):
```
Comment (anoto-moniz) [inline tests/test_session.py:297]: I'm not clear why we're removing this test.

File: src/citrine/citrine.py [hunk 7]
```diff
@@ -26,10 +26,24 @@ class Citrine:
...
-                 scheme: str = 'https',
+                 legacy_scheme: Optional[str] = None,
                  host: str = environ.get('CITRINE_API_HOST'),
-                 port: Optional[str] = None):
-        self.session: Session = Session(api_key, scheme, host, port)
+                 port: Optional[str] = None,
+                 *,
+                 scheme: str = 'https'):
+        if legacy_scheme is not None:
+            warn("Creating a session with positional arguments other than refresh_token "
```
[the remark this answers is not in the document]
Comment (kroenlein) [reply src/citrine/citrine.py:42]: Yes.  Copypasta following from Session changing the name of the argument.

File: tests/test_session.py (hunk 6 above)
Comment (kroenlein) [reply tests/test_session.py:297]: It's redundant with the tests above, and this test is written to do clever things depending on implementation.  For example, if you actually were to call `Citrine(host=None)`, you get an `InvalidURL` exception fired off from deeper in the stack, so under no circumstances would we care how it builds a URL with a null host.  But this test mocks the deeper calls, so this error is bypassed and a fatal scenario is treated as reasonable.

**Comment by kroenlein** [comment]
Raising an exception for a missing API key was breaking some tests.  Upon deeper digging, it made sense to remove that requirement because running with local devkit uses no API key and adding an extra conditional checking for `host=="localhost"` seemed potentially complicated.

File: tests/test_citrine.py [hunk 8]
```diff
@@ -22,14 +21,53 @@ def refresh_token(expiration: datetime = None) -> dict:
...
+        patched_key = "5678"
+        patched_host = "monkeypatch.citrine-testing.fake"
+        monkeypatch.setenv("CITRINE_API_KEY", patched_key)
+        monkeypatch.setenv("CITRINE_API_HOST", patched_host)
+        m.post(f'https://{patched_host}/api/v1/tokens/refresh', json=token_refresh_response)
+        m.get(f'https://{patched_host}/api/v1/utils/runtime-config', json=dict())
+
+        assert patched_key == Citrine().session.refresh_token
+        assert patched_key == Citrine(patched_key).session.refresh_token
+
```
Comment (anoto-moniz) [inline tests/test_citrine.py:52]: I think we should add a confirmation that ti works when the API key is omitted and CITRINE_API_KEY is not defined.

File: tests/test_session.py [hunk 9]
```diff
@@ -49,6 +48,48 @@ def session():
...
+        patched_key = "5678"
+        patched_host = "monkeypatch.citrine-testing.fake"
+        monkeypatch.setenv("CITRINE_API_KEY", patched_key)
+        monkeypatch.setenv("CITRINE_API_HOST", patched_host)
+        m.post(f'https://{patched_host}/api/v1/tokens/refresh', json=token_refresh_response)
+        m.get(f'https://{patched_host}/api/v1/utils/runtime-config', json=dict())
+
+        assert patched_key == Session().refresh_token
+        assert patched_key == Session(patched_key).refresh_token
+
```
Comment (anoto-moniz) [inline tests/test_session.py:73]: Same as above.

**Review by anoto-moniz — APPROVED** [review]
There's one small check I think we should add in the tests, but other than that looks great!

**Review by anoto-moniz — APPROVED** [review]
LGTM!

**Pull request merged**